### DISCUSSION TITLE
Dynamically update theme primary color

### DISCUSSION THREAD

**Discussion by gorgekara** [heroui-inc/heroui#782] [Help]
The idea is simple - once you create a theme (`createTheme`) you should be able to modify dynamically parts of it (for example - primary color) so its applied instantly to buttons, modals etc.

Not sure if this is part of nextui yet but if it isn't i believe it would be a good feature to have.

Example:

// _app.tsx
```
const theme = createTheme({
  theme: {
    colors: {
      primary: '#0088FF',
    }
  }
});

...

return (
   <NextUIProvider theme={theme}>
      <Component {...pageProps} />
   </NextUIProvider>
);
```

// ComponentA.tsx
```
import { changeTheme } from '@nextui-org/react';
const ComponentA = ({ dynamicColor }) => {
  React.useEffect(() => {
     changeTheme({
       colors: {
         primary: dynamicColor,
       }
     });
  }, []);

  return (
    <>
      ...next ui buttons, modals etc...
    </>
  );
};

Component.getInitialProps = async () => {
   const dynamicColor = await fetchDynamicColor();
   return { dynamicColor };
};
```

**Comment by shawnmclean** [comment]
Do we have any way of doing this now? I want to pull theme from database.

**Comment by fillobotto** [comment]
This is should get way more attention

**Comment by muneeb-r** [comment]
@shawnmclean how about you define themes statically in the code and referrence them into you DB and change dynamically?

**Comment by ChromeGG** [comment]
Bump, I also need this feature. For now, I'll try with workaround described by @muneeb-r , but if everything goes well and I'll get more customers, this solution will be unscalable

**Reply by muneeb-r** [reply]
@ChromeGG May be you can just utilize db and store theme in localstorage afterwards

**Reply by ChromeGG** [reply]
I have multi-tenant app, every tenant has custom domain. Storing themes in localstorage is not an option.

**Comment by lassschr** [comment]
Any news regarding this feature? ➕

**Reply by ChromeGG** [reply]
@lassschr I made workaround like that:

```tsx
// layout.tsx
import Color from 'color';
import { getPalette } from 'tailwindcss-palette-generator/getPalette';

export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) { 
  // API request
  const brandingColor = query.body.branding.ui.light.colors.primary;
  const palette = query.body.branding.ui.light.colors;

  const rawPalette = getPalette([{ name: 'primary', color: primarySeedColor }]);

  const primaryPalette = rawPalette['primary'] as Record<string, string>;
  const cssVariables: string[] = [];

  for (const shade in primaryPalette) {
    if (shade === 'DEFAULT') {
      continue;
    }
    const colorValue = primaryPalette[shade];
    const [h, s, l] = Color(colorValue).hsl().round().array();
    cssVariables.push(`--heroui-primary-${shade}: ${h} ${s}% ${l}%;`);
  }

  const defaultColor = Color(primaryPalette['DEFAULT']);
  const [h, s, l] = defaultColor.hsl().round().array();

  cssVariables.push(`--heroui-primary: ${h} ${s}% ${l}%;`);
  cssVariables.push(`--heroui-content1: ${palette['content1']};`);
  cssVariables.push(`--heroui-content2: ${palette['content2']};`);
  // ...
  cssVariables.push(`--heroui-divider: ${palette['divider']};`);

  cssVariables.push(`--heroui-focus: ${palette['focus']};`);

  return (
    <head>
          {/* ! inject css variables */}
          <style>{`:root { ${cssVariables.join('\n')} }`}</style>
          <link
            rel="apple-touch-icon"
            href="https://dev-assets.theplatform.pl/devstyle/favicon-192x192.png"
          />
        </head>
)

```

With a little help of LLM, it's possible to achieve 😄

**Reply by lassschr** [reply]
Hi, thank you for your comment! I have been having something similar as a backup plan 😁 Was hoping to avoid manually overwriting styles if possible - but thank you for sharing 🥇

**Reply by umutnacak** [reply]
@ChromeGG Hi, have you tried changing variables other than colors? I tried it for demo purposes (--spacing), but I wonder if it will cause problems in the future. LLM tells me to avoid it if possible, but when I give my reason, it shows this solution.

**Reply by ChromeGG** [reply]
@umutnacak Hmm, I haven't tried with other properties than colors. But I believe it should behave in the same way, they are just CSS variables under the hood as colors are

**Reply by umutnacak** [reply]
@ChromeGG Thanks for the reply. Yes, but one needs to be careful. Since I'm not a designer, the UI breaks down very quickly and I don't know what I'm doing. Thanks again and good luck with your app.

**Comment by baiwumm** [comment]
https://github.com/user-attachments/assets/24bfe7c0-7514-4e9e-a277-8a35b5db449a

This is my current implementation plan, you have maybe you can refer to: https://next.baiwumm.com/
